packages/waku/src/lib/plugins/vite-plugin-typegen.ts

```diff
@@ -0,0 +1,146 @@
+const OUTPUT_FILE = path.resolve(process.cwd(), './entries.gen.tsx');
+
+// Recursively collect `.tsx` files in the given directory
+const collectFiles = async (dir: string): Promise<string[]> => {
+  let results: string[] = [];
+  const files = await promises.readdir(dir, { withFileTypes: true });
+
+  for (const file of files) {
+    const fullPath = path.join(dir, file.name);
+    if (file.isDirectory()) {
+      results = results.concat(await collectFiles(fullPath));
+    } else if (file.isFile() && fullPath.endsWith('.tsx')) {
+      results.push(fullPath.replace(SRC_DIR, ''));
+    }
+  }
+  return results;
+};
+
+const srcToName = (src: string) => {
+  const split = src
+    .split('/')
+    .slice(0, -1)
+    .filter(Boolean) // ?
+    .map((part) => part[0]!.toUpperCase() + part.slice(1));
+
+  if (src.endsWith('_layout.tsx')) {
+    return split.join('') + 'Layout';
+  } else if (src.endsWith('index.tsx')) {
+    return split.join('') + 'Index';
+  } else {
+    const fileName = src.split('/').at(-1)!.replace('.tsx', '');
+    return split.join('') + fileName[0]!.toUpperCase() + fileName.slice(1);
+  }
+};
+
+const fileExportsGetConfig = (filePath: string) => {
+  const file = readFileSync(SRC_DIR + filePath).toString();
+
+  return (
+    file.includes('const getConfig') || file.includes('function getConfig')
+  );
+};
+
+const generateFile = async (filePaths: string[]) => {
+  const fileInfo = [];
+  for (const filePath of filePaths) {
+    // where to import the component from
+    const src = filePath.slice(1);
+    const hasGetConfig = fileExportsGetConfig(filePath);
+
+    if (filePath.endsWith('_layout.tsx')) {
+      fileInfo.push({
+        type: 'layout',
+        path: filePath.replace('_layout.tsx', ''),
+        src,
+        hasGetConfig,
+      });
+    } else if (filePath.endsWith('index.tsx')) {
+      fileInfo.push({
+        type: 'page',
+        path: filePath.replace('index.tsx', ''),
+        src,
+        hasGetConfig,
+      });
+    } else {
+      fileInfo.push({
+        type: 'page',
+        path: filePath.replace('.tsx', ''),
+        src,
+        hasGetConfig,
+      });
+    }
+  }
+
+  let result = `import { createPages } from 'waku';
+import type { PathsForPages } from 'waku/router';\n\n`;
+
+  for (const file of fileInfo) {
+    const moduleName = srcToName(file.src);
+    result += `import ${moduleName}${file.hasGetConfig ? `, { getConfig as ${moduleName}_getConfig }` : ''} from './src/pages/${file.src.replace('.tsx', '')}';\n`;
+  }
+
+  result += `\nconst DEFAULT_CONFIG = {render: 'dynamic'} as const;
+  
+  const _pages = createPages(async (pagesFns) => [\n`;
+
+  for (const file of fileInfo) {
+    const moduleName = srcToName(file.src);
+    result += `pagesFns.${file.type === 'layout' ? 'createLayout' : 'createPage'}({ path: '${file.path}', component: ${moduleName}, ...${file.hasGetConfig ? `(await ${moduleName}_getConfig())` : 'DEFAULT_CONFIG'} }),\n`;
```

Comment:
@dai-shi Is this line hard to read? If yes, do you have any recommendations for simplifying it?

Reply:
there might be room for improvement, but fine for now.